server/src/com/thoughtworks/go/server/service/EnvironmentConfigService.java

```diff
@@ -223,25 +225,157 @@ private void editEnvironments(EditEnvironments edit, EnvironmentConfig config, U
+        private final String uuid;
+        private final String environmentName;
+        private final TriStateSelection.Action action;
+
+        public ModifyEnvironmentCommand(String uuid, String environmentName, TriStateSelection.Action action) {
+            this.uuid = uuid;
+            this.environmentName = environmentName;
+            this.action = action;
+        }
+
+        public CruiseConfig update(CruiseConfig cruiseConfig) throws Exception {
+            AgentConfig agentConfig = cruiseConfig.agents().getAgentByUuid(uuid);
+            bombIfNull(agentConfig, "Unable to set agent resources; Agent [" + uuid + "] not found.");
+            EnvironmentConfig environmentConfig = cruiseConfig.getEnvironments().find(new CaseInsensitiveString(environmentName));
+            if (environmentConfig == null) {
+                agentConfig.addError("environments", "Environment [" + environmentName + "] does not exist.");
+            } else {
+                if (action.equals(TriStateSelection.Action.add)) {
+                    environmentConfig.addAgentIfNew(uuid);
+                } else if (action.equals(TriStateSelection.Action.remove)) {
+                    environmentConfig.removeAgent(uuid);
+                } else if (action.equals(TriStateSelection.Action.nochange)) {
+                    //do nothing
+                } else {
+                    bomb(String.format("unsupported action '%s'", action));
+                }
+            }
+            return cruiseConfig;
+        }
+    }
+
+    public static class DeleteEnvironmentCommand implements UpdateConfigCommand, UserAware {
+        private EnvironmentConfig environmentConfig;
+        private Username username;
+
+        public DeleteEnvironmentCommand(EnvironmentConfig environmentConfig, Username username) {
+            this.environmentConfig = environmentConfig;
+            this.username = username;
+        }
+
+        @Override
+        public CruiseConfig update(CruiseConfig cruiseConfig) throws Exception {
+            cruiseConfig.getEnvironments().remove(environmentConfig);
+            return cruiseConfig;
+        }
+
+        @Override
+        public ConfigModifyingUser user() {
+            return new ConfigModifyingUser(username);
+        }
+    }
+
+    public static class AddEnvironmentCommand implements UpdateConfigCommand, UserAware {
+        private final BasicEnvironmentConfig environmentConfig;
+        private Username user;
+
+        public AddEnvironmentCommand(BasicEnvironmentConfig environmentConfig, Username user) {
+            this.environmentConfig = environmentConfig;
+            this.user = user;
+        }
+
+        public CruiseConfig update(CruiseConfig cruiseConfig) {
+            cruiseConfig.addEnvironment(environmentConfig);
+            return cruiseConfig;
+        }
+
+        @Override
+        public ConfigModifyingUser user() {
+            return new ConfigModifyingUser(user);
+        }
+    }
```

Comment:
Consider moving AddEnvironmentCommand, DeleteEnvironmentCommand, UpdateEnvironmentCommand, ModifyEnvironmentCommand out of the service for readability